File: src/main/java/frc/robot/subsystems/superstructure/SuperstructureState.java
package frc.robot.subsystems.superstructure;

import com.pathplanner.lib.path.PathConstraints;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.util.struct.Struct;
import edu.wpi.first.util.struct.StructSerializable;
import frc.robot.subsystems.swerve.Drivebase;
import java.nio.ByteBuffer;
import java.util.Optional;

/** Immutable description of a superstructure target state. */
public final class SuperstructureState implements StructSerializable {
  public static final Struct<SuperstructureState> STRUCT = new SuperstructureStateStruct();

  private final Rotation2d armAngle;
  private final PathConstraints driveLimits;
  private final Optional<SuperstructureState> nextState;
  private final String name;

  public SuperstructureState(Rotation2d armAngle) {
    this(armAngle, Drivebase.getDefaultLimits(), Optional.empty(), "");
  }

  public SuperstructureState(
      Rotation2d armAngle,
      PathConstraints driveLimits,
      Optional<SuperstructureState> nextState,
      String name) {
    this.armAngle = armAngle;
    this.driveLimits = driveLimits;
    this.nextState = nextState == null ? Optional.empty() : nextState;
    this.name = name;
  }

  public Rotation2d getArmAngle() {
    return armAngle;
  }

  public PathConstraints getDriveLimits() {
    return driveLimits;
  }

  public Optional<SuperstructureState> getNextState() {
    return nextState;
  }

  public String getName() {
    return name;
  }

  @Override
  public Struct<SuperstructureState> getStruct() {
    return STRUCT;
  }

  private static final class SuperstructureStateStruct implements Struct<SuperstructureState> {
    @Override
    public Class<SuperstructureState> getTypeClass() {
      return SuperstructureState.class;
    }

    @Override
    public String getTypeName() {
      return "SuperstructureState";
    }

    @Override
    public int getSize() {
      return Double.BYTES;
    }

    @Override
    public String getSchema() {
      return "double armAngle;";
    }

    @Override
    public void pack(ByteBuffer bb, SuperstructureState value) {
      bb.putDouble(value.armAngle.getRadians());
    }

    @Override
    public SuperstructureState unpack(ByteBuffer bb) {
      double angleRad = bb.getDouble();
      return new SuperstructureState(Rotation2d.fromRadians(angleRad));
    }
  }
}

File: src/main/java/frc/robot/subsystems/superstructure/SuperstructureState.kt
package frc.robot.subsystems.superstructure

import com.pathplanner.lib.path.PathConstraints
import edu.wpi.first.units.Units.Meters
import edu.wpi.first.units.Units.Radians
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.Distance
import edu.wpi.first.util.struct.Struct
import edu.wpi.first.util.struct.StructSerializable
import frc.robot.subsystems.swerve.Drivebase
import lib.math.units.into
import java.nio.ByteBuffer
import java.util.*

data class SuperstructureState(
    val armAngle: Angle,
    val driveLimits: PathConstraints = Drivebase.defaultLimits,
    val nextState: Optional<SuperstructureState> = Optional.empty(),
    val name: String = ""
) : StructSerializable {
    companion object {
        class SuperstructureStateStruct : Struct<SuperstructureState> {
            override fun getTypeClass(): Class<SuperstructureState> {
                return SuperstructureState::class.java
            }

            override fun getTypeName(): String {
                return "SuperstructureState"
            }

            override fun getSize(): Int {
                return Double.SIZE_BYTES
            }

            override fun getSchema(): String {
                return "double armAngle;"
            }

            override fun pack(bb: ByteBuffer?, value: SuperstructureState?) {
                bb!!.putDouble(value!!.armAngle into Radians)
            }

            override fun unpack(bb: ByteBuffer?): SuperstructureState {
                return SuperstructureState(
                    Radians.of(bb!!.getDouble()),
                )
            }
        }

        val struct = SuperstructureStateStruct()
    }
}